cli/lib/commands/help.js

```diff
@@ -1,44 +1,67 @@
 'use strict';
 
-const fs = require('fs');
-const path = require('path');
-const { promisify } = require('util');
-const readdirAsync = promisify(fs.readdir);
+const helpFile = require('./help.json');
+const chalk = require('chalk');
 
 const userHome = require('user-home');
 
 module.exports = help;
 
-async function help(opts) {
+function help(opts) {
   const command = opts.argv[0];
   if (!command) {
-    await showBasicHelp();
+    showBasicHelp();
   } else {
-    return new Promise((resolve, reject) => {
-      fs.createReadStream(path.join(__dirname, `help-${command}.txt`))
-        .on('error', async err => {
-          if (err.code === 'ENOENT') {
-            console.log(
-              `help has not been implemented yet for ${command}. You could build it!`
-            );
-            await showBasicHelp();
-            return resolve();
-          }
-          reject(err);
-        })
-        .on('end', () => resolve())
-        .pipe(process.stdout);
-    });
+    if (helpFile[command] == null) {
+      console.log(
+        `help has not been implemented yet for ${command}. You could build it!`
+      );
+      showBasicHelp();
+    } else {
+      const text = helpFile[command];
+      let first = true;
+      for (let line of text) {
+        if (first) {
+          console.log('\nHelp for: ' + parse(line) + '\n');
+          first = false;
+        } else {
+          console.log('\t' + parse(line));
+        }
+      }
+    }
   }
 }
 
-async function showBasicHelp() {
-  const commands = (await readdirAsync(__dirname))
-    .filter(cmd => cmd.endsWith('.js'))
-    .map(cmd => `\t${cmd.split('.')[0]}`)
-    .join('\n');
+const parse = sArr => {
```

Comment:
As someone who primarily learned programming through FP/FP-lite kinda concepts, this is a little hard for me to read. Would it be worth adding some comments here for other folks who have a hard time visually parsing nested loops and mutation?

Reply:
Actually looking over this, I wonder if it would make more sense to just not have styles in the descriptions of commands/in the actual help text. That would make the styling code much easier to deal with (style the keys, don't style the help text), which would in turn make this easier to maintain and extend in the future, including adding translations. What do you think?

Reply:
This function also makes using characters that aren't in the basic multi-lingual plane super hard. [Mathias Bynens, as usual, [has great details on this](https://mathiasbynens.be/notes/javascript-encoding)]

The following program:

````js
var s = '🤖👻💃'
for (let i = 0; i < s.length; i++) { console.log(s[i])}
```

Will output a ton of mojibake:

```
�
�
�
�
�
�
```

This is because javascript treats the code unit as individual characters, but for characters outside the BMP, this isn't the case, and the "character" itself (as a human would read it) can consist of up to four "characters" as javascript would count it.

So if we were to translate a command into a language that uses characters outside the BMP, this would count wrong for those translations.

Reply:
> Actually looking over this, I wonder if it would make more sense to just not have styles in the descriptions of commands/in the actual help text. That would make the styling code much easier to deal with (style the keys, don't style the help text), which would in turn make this easier to maintain and extend in the future, including adding translations. What do you think?

I think we can get a little bit of both to be honest! I think color is a powerful tool to help show the command names/keywords, but I think we can get there with a transform stream.  I should have something to show this week but ideally you'd do something like:

```js
const HighlightStream = require('highlight-stream')
const keywords = ['ds install', 'ds build']
const highlighter = new HighlitStream(keywords)
fs.createReadStream('helpfile').pipe(highlighter).pipe(process.stdout)
```

And `ds install` and `ds build` would be highlighted, but the `helpfile` is just a text file with no formatting commands.

Reply:
(There are also library like https://www.npmjs.com/package/unicount for dealing with javascript's horrible handling of "big" characters)